### ISSUE TITLE
^\underline and greediness

### ISSUE THREAD

**Issue by HUSTER-Qin** [KaTeX/KaTeX#2046]
```latex
error 
\text{Cu + 2}{{\text{H}}_2}\text{S}{{\text{O}}_4}~^\underline{\underline{\Delta}}~\text{CuS}{{\text{O}}_4}\text{ + S}{{\text{O}}_2}\uparrow + \ 2{{\text{H}}_2}\text{O}
ok
\text{Cu + 2}{{\text{H}}_2}\text{S}{{\text{O}}_4}~^{\underline{\underline{\Delta}}}~\text{CuS}{{\text{O}}_4}\text{ + S}{{\text{O}}_2}\uparrow + \ 2{{\text{H}}_2}\text{O}
problem:
~^ => ~^{\underline{\underline{\Delta}}}  why?
How to use macros to solve this problem？
```

**Comment by edemaine** [comment]
I confirmed that `x^\underline{y}` works in LaTeX.  Looking at the KaTeX code, I believe this bug is caused by `\underline`'s `greediness` being set too low.  Sup/subscripts `^`/`_` have greediness `1` (via `SUPSUB_GREEDINESS`), while `\underline` has no greediness set, which defaults to `1`.

Is greediness a TeX notion?  If so, there should be reference we can follow.  Either way, I think we need to review all the greediness settings... or possibly just change the default to be greater than `SUPSUB_GREEDINESS`.

**Comment by edemaine** [comment]
@HUSTER-Qin Regarding your "how to solve with macros", a workaround would be to define a macro such as `\u` to `{\underline{#1}}`, and then use `\u` instead of `\underline`.

**Comment by edemaine** [comment]
I did some more experiments, this time in `pdflatex` instead of QuickLaTeX (which sometimes doesn't report errors).  I found that `x^\frac12`, `x^\frac{x^2}{y^2}`, `x^\mathrm h`, `x^\mathrm{hi}`, `x^\text h`, and `x^\text{hi}` all worked (everything after `^` rendered in the subscript), while `x^\sqrt 5`, `x^\underline y`, `x^\bar y`, and `x^\underline x` all return errors (though TeX usually recovers correctly and renders what you might have wanted, sometimes it does weird things).  So I'm no longer convinced this issue is a bug.

Incidentally, `\sqrt \frac 1 2` also fails, whereas it is described in KaTeX's `defineFunction.js` as explicitly allowed via greediness (for MathJax compatibility?).

I would definitely find setting `SUPSUB_GREEDINESS` to `0` more convenient to use; it would make all of the examples above work.  But it would not be LaTeX compatible, so a user would be in for a surprise if trying to copy their working KaTeX inputs over to LaTeX.

What distinguishes `\frac`, `\mathrm`, and `\text` that make them work in an unbraced subscript?  I think `\frac` is special because it expands to `{{#1} \over #2}`, so in particular it's wrapped in braces.  `\text` expands to a `\@text` call which is also wrapped in braces (`{\mathchoice ...}`).  It's much harder to read the definition of `\mathrm`, but this seems to be the general rule: an unbraced argument works as a sub/superscript if and only if it expands to a single token or a braced group of tokens.  This is how undelimited arguments work too in TeX, so that makes sense.

Perhaps this explains why `\mathrm`, `\frac`, and `\text` all have greediness 2 already in KaTeX.  As a result, all the cases that I've found to work in LaTeX also work in KaTeX.  So maybe we're done?  On the other hand, `\textcolor` and `\colorbox` have greediness 3, which means that `x^\textcolor{blue}y` works in KaTeX (but not LaTeX); and lots of `\frac` lookalikes in `functions/genfrac.js`, e.g. `\cfrac`, have greediness 2 so `x^\cfrac12` works in KaTeX but I tested and it doesn't work in LaTeX.  Do we care enough about this incompatibilities?
